Adding ILogger abstaction and replacing LogEventSource with it

ricardSiliuk opened the pull request:
Sorry for a new PR, messed something up locally 😢 

There are a few TODOs which are essentially requests for comments from you. I'm not sure whether rewriting `Client.cs` to become `ClientFactory.cs` should be in scope of this PR but it would allow me to replace `Debug.WriteLine` with `_logger.LogDebug()` (it's static so I can't use the logger instance there 😢). Same with `CreditResponse.cs` and one other file.

Also wording of some messages can be redone - essentially, upon creation of object logger could create scope with `Identity`, `StreamIdentifier` and `ClientType` (`Consumer`, `Producer`, etc.) already filled out and then we don't need to log those as part of the message - it'll just exist in scope (sort of like MDC in Java?). Not sure whether that's desirable from your side so let me know.

Also, as part of refactoring moved constructors to the top of class for easier navigation.

Also, I commited a test file to see how creation of clients looks like when invoked by hand - that should be removed when we're ready to merge.

Gsantomaggio commented:
closed in favour of https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/pull/156

Gsantomaggio commented:
@ricardSiliuk It is not clear to me how to pass the logger on the `IReconnectStrategy`. For example:

```csharp
       
var serviceCollection = new ServiceCollection();
serviceCollection.AddLogging(builder => builder
    .AddConsole()
    .AddFilter(level => level >= LogLevel.Trace)
);
var loggerFactory = serviceCollection.BuildServiceProvider()
    .GetService<ILoggerFactory>();


if (loggerFactory != null)
{
    var producerLogger = loggerFactory.CreateLogger<Producer>();
    var consumerLogger = loggerFactory.CreateLogger<Consumer>();
 .....

var producer = await Producer.Create(
                new ProducerConfig(system, stream), producerLogger);
                
```


But I can't see the `IReconnectStrategy` logs. Should we pass the producer/consumer reference to the `IReconnectStrategy` ?

Gsantomaggio reviewed (changes requested):
please read https://github.com/rabbitmq/rabbitmq-stream-dotnet-client/pull/190#issuecomment-1346313498 

Since we are here, I'd like to add some debug log to the `StreamSystem` class
Can you please add also the `_logger` there?

ricsiLT commented:
Yes, I also had my doubts as to whether `IReconnectStrategy` can have a logger... Technically, you can just pass in the one consumer/producer has? It wouldn't be categorized separately (would just be `ILogger` instead of `ILogger<XReconnectStrategy>`) but truth be told - it probably doesn't have to.

ricsiLT commented:
Re: logger in stream system - sure, will add :)

ricardSiliuk commented:
@Gsantomaggio I think it's best to move strategy into `Producer`/`Consumer` directly instead of passing it as part of configuration. There we either inject separate logger, or reuse the one passed to `Producer`/`Consumer` :)

Gsantomaggio commented:
> Technically, you can just pass in the one consumer/producer has? 

yes I would use the one you configured for the `Producer`/`Consumer`

ricardSiliuk commented:
You mean, for example, in here?

```
    // <summary>
    // Create a new Producer
    // </summary> 
    public static async Task<Producer> Create(ProducerConfig producerConfig, ILogger<Producer> logger = null)
    {
        var rProducer = new Producer(producerConfig, logger);
        await rProducer.Init(producerConfig.ReconnectStrategy);
        return rProducer;
    }
```

I think that's not possible, right? `ProducerConfig` is already created here, so `ReconnectionStrategy` already defaulted to it's value. Maybe `ReliableConfig` should not default to anything, and lets put responsibility of it to `Create()` methods? Then we could check whether `XConfig.ReconnectStrategy` is null and pass in a logger to a `Strategy` we'd create on the spot.

But at that point we could adjust this `Create()` method to take a strategy separately from config and that strategy can have a logger if it so desires.

Which way should I go?

Gsantomaggio commented:
Something like:
```csharp
    public static async Task<Producer> Create(ProducerConfig producerConfig, ILogger<Producer> logger = null)
    {
        producerConfig.ReconnectStrategy ??= new BackOffReconnectStrategy(logger);
```

and remove the initialization form:
```csharp
public record ReliableConfig
{
    public IReconnectStrategy ReconnectStrategy { get; set; } = new BackOffReconnectStrategy()
```

ricardSiliuk commented:
Yeah, makes sense! Will add shortly :)

ricardSiliuk commented:
Added strategy as part of `Config` initialized via `Consumer`/`Producer` though it still feels better stored elsewhere. In `ReliableBase` maybe? 

Otherwise, almost ready for merge. There's some TODOs I'd like you to review and then we're golden.

Gsantomaggio commented:
>In ReliableBase maybe?

why not, if we can remove code duplication

ricardSiliuk commented:
Moved strategy to `ReliableBase`. Looks quite nice. Also added some comments/cleaned up existing ones.
As for `LogEvent`... Yeah, it should probably be removed.

Gsantomaggio commented:
The basic idea about `*Config config` is to have only one config as reference. 
This one:
`public async Task<Consumer> Create(ConsumerConfig config, IReconnectStrategy reconnectStrategy, ILogger<Consumer> logger);` is too much.


I have a proposal:

Instead of passing all these `ILogger` around all the classes. Can we just have "one" `ILogger` and use it everywhere?
I mean, when we create the `StreamSystem`, we define the `ILogger`, and we can use it.  That's it.

More or less how we do with ` public static class StreamCompressionCodecs`

ricsiLT commented:
I'll look around tomorrow but that also doesn't sound right. At least from the codebases I work with :D I know `Serilog` has a single static logger if you want it, and `Log4Net` probably has a similar one. 

```
public async Task<Consumer> Create(ConsumerConfig config, IReconnectStrategy reconnectStrategy, ILogger<Consumer> logger);
```

Is this too much because it has too many arguments, or are you worried that more will come in the future? They could default to null but then you either call this func with named args or end up with a lot of nulls :D 

```
var logger = loggerFactory.CreateLogger<Consumer>();

var consumer = Consumer.Create(config, null, null);
var consumer_2 = Consumer.Create(config, logger: logger);

etc...
```

Gsantomaggio commented:
With this PR we introduce a new and flexible way to implement the logs.
How to test:
```csharp
 public static async Task Start()
    {
        var serviceCollection = new ServiceCollection();
        serviceCollection.AddLogging(builder => builder
            .AddSimpleConsole(options =>
            {
                options.IncludeScopes = true;
                options.SingleLine = true;
                options.TimestampFormat = "[HH:mm:ss] ";
                options.ColorBehavior = LoggerColorBehavior.Default;
            })
            .AddFilter(level => level >= LogLevel.Information)
        );
        var loggerFactory = serviceCollection.BuildServiceProvider()
            .GetService<ILoggerFactory>();

        if (loggerFactory != null)
        {
            var producerLogger = loggerFactory.CreateLogger<Producer>();
            var consumerLogger = loggerFactory.CreateLogger<Consumer>();
            var streamLogger = loggerFactory.CreateLogger<StreamSystem>();


            var config = new StreamSystemConfig
            {
                UserName = "guest",
                Password = "guest",
                VirtualHost = "/"
            };
            var system = await StreamSystem.Create(config, streamLogger);

            // Create the stream. It is important to put some retention policy 
            await system.CreateStream(new StreamSpec(stream)
            {
                MaxLengthBytes = 2_000_000_000_000,
            });

            var producer = await Producer.Create(
                new ProducerConfig(system, stream)
                {
                    Reference = "my_prod"
                }, producerLogger);

            // Publish the messages
            for (var i = 0; i < 2; i++)
            {
                var message = new Message(Encoding.UTF8.GetBytes($"A{i}"));
                {
                 
                };

              
                await producer.Send(message);
            }

// Create a consumer
            var consumer = await Consumer.Create(
                new ConsumerConfig(system, stream)
                {
                    Reference = "my_consumer",
                    // Consume the stream from the beginning 
                    // See also other OffsetSpec 
                    OffsetSpec = new OffsetTypeFirst(),
                    // Receive the messages
                    MessageHandler = async (sourceStream, consumer, ctx, message) =>
                    {
                        Console.WriteLine(
                            $"message: coming from {sourceStream} data: {Encoding.Default.GetString(message.Data.Contents.ToArray())} - consumed");
                        await Task.CompletedTask;
                    }
                }, consumerLogger);


            var rawProducerLogger = loggerFactory.CreateLogger<RawProducer>();
            var rawConsumerLogger = loggerFactory.CreateLogger<RawConsumer>();
          
            var rawProducer = await system.CreateRawProducer(new RawProducerConfig(stream), rawProducerLogger);
            var rawConsumer = await system.CreateRawConsumer(new RawConsumerConfig(stream), rawConsumerLogger);
          
            var rawSuperProducerLogger = loggerFactory.CreateLogger<RawSuperStreamProducer>();
            var rawSuperConsumerLogger = loggerFactory.CreateLogger<SuperStreamConsumer>();
            var rawSuperProducer = await system.CreateRawSuperStreamProducer(new RawSuperStreamProducerConfig("invoices")
            {
                Routing = message => "1"
            }, rawSuperProducerLogger);
            var rawSuperConsumer = await system.CreateSuperStreamConsumer(new SuperStreamConsumerConfig("invoices"), rawSuperConsumerLogger);



            Console.WriteLine($"Press to stop");
            var x = Console.ReadLine();

            await producer.Close();
            await consumer.Close();
            await system.DeleteStream(stream);
            await system.Close();
            Console.WriteLine($"Press to stop delete");
            Console.ReadLine();

        }
    }
```


each class has its own logger, most of the users will use:
```csharp
       var producerLogger = loggerFactory.CreateLogger<Producer>();
       var consumerLogger = loggerFactory.CreateLogger<Consumer>();
       var streamLogger = loggerFactory.CreateLogger<StreamSystem>();
```

The interesting part is that we can decide to use different loggers for different contexts, for example:

```csharp
     var producerWithLogger = await Producer.Create(
                new ProducerConfig(system, stream)
                {
                    Reference = "my_prod"
                }, producerLogger);

     var producerNoLogger = await Producer.Create(
                new ProducerConfig(system, stream)
                {
                    Reference = "my_prod_no"
                });


```

lukebakken commented on RabbitMQ.Stream.Client/Message.cs:
```diff
@@ -131,7 +131,6 @@ public static Message From(ref SequenceReader<byte> reader, uint len)
                         offset += AmqpWireFormatting.ReadAny(ref reader, out amqpValue);
                         break;
                     default:
-                        LogEventSource.Log.LogError($"dataCode: {dataCode} not handled. Please open an issue.");
```
I have seen several times where a very-unexpected exception throws a message that includes a link to report an issue as well as important information to include with the issue. In this case it would be client library version, RabbitMQ version (if we have it), etc ... is that worth doing here or in another PR?

Gsantomaggio replied:
Maybe another PR

Gsantomaggio commented:
Thank you @ricardSiliuk
